Documentation for the Logging stdlib

c42f opened the pull request:
Here's some documentation covering the Logging stdlib.

Fixes #25154

c42f committed:
Documentation for the Logging stdlib

stevengj commented on stdlib/Logging/docs/src/index.md:
```diff
@@ -0,0 +1,202 @@
+enabled.
+
+Second, the logging tools allow you to attach arbitrary data to each event as a
+set of key value pairs. This allows you to capture local variables and other
```
`key=value` pairs

stevengj commented on stdlib/Logging/docs/src/index.md:
```diff
@@ -0,0 +1,202 @@
+    early return is more appropriate.)
+* The *message* is a human readable string describing the event in markdown
+  format.
+* Optional *key value pairs* allow arbitrary data to be attached to each event.
```
key–value pairs

stevengj commented on stdlib/Logging/docs/src/index.md:
```diff
@@ -0,0 +1,202 @@
+[`AbstractLogger`](@ref).
+
+When an event is triggered, the appropriate logger is found by looking for a
+task local logger with the global logger as fallback.  The idea here is that
```
task-local logger

(hyphen, not an en dash)

stevengj commented on stdlib/Logging/docs/src/index.md:
```diff
@@ -0,0 +1,202 @@
+task local logger with the global logger as fallback.  The idea here is that
+the application code knows how log events should be processed and exists
+somewhere at the top of the call stack. So we should look up through the call
+stack to discover the logger - that is, the logger should be *dynamically
```
em dash (—), not hyphen

(I end up giving all my students the talk about hyphen vs en dash vs em dash vs minus sign. 😉 )

c42f replied:
Thanks. I normally get the emdash correct, but there's no nice way to type it in markdown so I got sloppy ;-) Oddly enough, the julia markdown parser accepts the LaTeX `--` for endash but not `---` for emdash.

stevengj replied:
On most operating systems there is a keyboard shortcut to type the Unicode dash directly.  For example, on MacOS you type option-hyphen for en dash and option-shift-hyphen for em dash.

stevengj commented on stdlib/Logging/docs/src/index.md:
```diff
@@ -0,0 +1,202 @@
+
+Exceptions which occur while generating the log event are captured and logged
+by default.  This prevents individual broken events from crashing the
+application which is helpful when enabling little-used debug events in a
```
comma before "which"

stevengj commented on stdlib/Logging/docs/src/index.md:
```diff
@@ -0,0 +1,202 @@
+formatting and filtering.  [`NullLogger`](@ref) is a convenient way to drop all
+messages where necessary; it is the logging equivalent of the [`DevNull`](@ref)
+stream.  [`SimpleLogger`](@ref) is a very simplistic text formatting logger,
+mainly useful for debugging the logging system itself.
```
What methods should a custom logger define?

stevengj replied:
If these are the ones in the `Processing events` section below, you can just cross-reference that here.

c42f replied:
I've added a section about this.

stevengj commented on stdlib/Logging/docs/src/index.md:
```diff
@@ -0,0 +1,202 @@
+* A `group` for the event which is set to the base name of the file by default,
+  without extension.  This can be used to group messages into categories more
+  finely than the log level (for example, all deprecation warnings have group
+  `:depwarn`), or into logical groupings across or within modules.
```
Is there any way to attach additional information to log events?

For example, in IJulia we might want to attach additional information about the current execution cell, so that the log message is printed in the right output cell.   (If necessary, we can extract this from the `file` metadata, since the input cell name is used as the filename, but I'm curious if there is a way to attach additional runtime information.  For example, if you define a custom logger, could you implement a method to generate additional key-value pairs based on some dynamic global state at the time the log message is emitted?)

c42f replied:
Absolutely — dynamic context is something which can be added by installing a custom logger as a filter (add a key–value pair and pass to the next logger up the stack).  This could include information such as:
* Global or task-local variables as you're suggesting.
* The time of the event
* A backtrace

I think this is going to be a common enough pattern that we could support it in the library. For now however it's pretty bare bones and provides just enough structure for this kind of stuff to be done outside Base/stdlib.

For IJulia I expect you'll want to install a custom logger anyway because formatting of log messages can be a lot nicer in the browser compared to the console.

swt30 commented on stdlib/Logging/docs/src/index.md:
```diff
@@ -0,0 +1,202 @@
+
+* The *log level* is a broad category for the message which is used for early
+  filtering. There are several standard levels of type [`LogLevel`](@ref);
+  user defined levels are also possible.
```
user-defined

swt30 commented on stdlib/Logging/docs/src/index.md:
```diff
@@ -0,0 +1,202 @@
+* The *log level* is a broad category for the message which is used for early
+  filtering. There are several standard levels of type [`LogLevel`](@ref);
+  user defined levels are also possible.
+  - Use `Debug` for verbose information which could be useful when debugging an
```
I'd use "that" instead of "which" here.

c42f replied:
Aha, I've already removed most of these "which"s locally.  Will push an update shortly.

swt30 replied:
Excellent, you've clearly got your pronouns under control then ;)

swt30 commented on stdlib/Logging/docs/src/index.md:
```diff
@@ -0,0 +1,202 @@
+
+As you can see in the examples, logging statements make no mention of
+where log events go or how they are processed. This is a key design feature
+which makes the system composable and natural for concurrent use. It does this
```
I'd also use "that" instead of "which" here.

swt30 commented on stdlib/Logging/docs/src/index.md:
```diff
@@ -0,0 +1,202 @@
+
+### Loggers
+
+Processing of events is performed by a *logger* which is the first piece of
```
Comma before "which"

swt30 commented on stdlib/Logging/docs/src/index.md:
```diff
@@ -0,0 +1,202 @@
+composing functionality from multiple modules.)
+
+The global logger may be set with [`global_logger`](@ref), and task local
+loggers controlled using [`with_logger`](@ref).  Newly spawned tasks inherit
```
Newly-spawned

c42f replied:
Thanks, any gifts of compound adjective detection are gratefully received :-)

swt30 replied:
task-local in the line above too!

stevengj replied:
Actually, this is incorrect: you *don't* hyphenate compound adjectives where the modifier is an adverb ending with "ly":  https://www.merriam-webster.com/words-at-play/6-common-hypercorrections-and-how-to-avoid-them/hyphenating-ly-adverbs

swt30 replied:
^ that's true, my bad!

swt30 commented on stdlib/Logging/docs/src/index.md:
```diff
@@ -0,0 +1,202 @@
+somewhere at the top of the call stack. So we should look up through the call
+stack to discover the logger - that is, the logger should be *dynamically
+scoped*. (This is a point of contrast with logging frameworks where the
+logger is *lexically scoped* - provided explicitly by the module author or a
```
If you're fixing the hyphen -> em dash above you should do this one too, or you could just turn it into a comma: "where the logger is *lexically scoped*, provided explicitly by the module author"

c42f replied:
Already done (locally) :-)

swt30 reviewed (commented):
At the risk of turning this into a proofreading session, I popped in some more minor grammatical/punctuation suggestions along the same lines as stevengj. I really love the new logging system, by the way---well done!

c42f committed:
Grammar tweaks and other small fixes

Remove default implementations for shouldlog and min_enabled_level as
these seem more likely to cause confusion than be useful.

c42f commented:
I've pushed some tweaks to address the questions which came up here (thanks).

Writing the section about the AbstractLogger functions made me reconsider which functions should have default implementations.  I think the defaults for `shouldlog` and `min_enabled_level` were more likely to cause confusion than be useful so I've removed those for now.

quinnj commented on base/logging.jl:
```diff
@@ -169,9 +169,6 @@ overridden:
 
 There's also some key value pairs which have conventional meaning:
 
-  * `progress=fraction` should be used to indicate progress through an
```
why take this out?

c42f replied:
Yes, I loved this feature and removing it makes me sad.  But I think it's too unreliable as a user facing interface for progress reporting.  Consider

```julia
for i=1:N
    foo()
    @info "blah" progress=i/N
end
```

This looks super simple and nice, but what if `foo()` throws?  Then you're left with an incomplete progress bar which needs to be cleaned up with a heuristic of some kind.

We might "fix" this by limiting things to one progress bar at a time. But this doesn't make sense for nested progress bars and more general concurrent situations where the loop above may be executing from more than one task at once.  In these cases you really want a reliable way to detect when the progress step is complete.

So I think it's better to remove this for now and reintroduce a more reliable progress reporting interface later.

StefanKarpinski replied:
I think that's a sane approach. Better to be a little cautious now and give the good stuff later.

quinnj reviewed (commented):
Love it! My only request would be maybe include a very quick "example AbstractLogger"? Maybe something like a "JSONLogger" that output everything in JSON? Just a thought.

stevengj commented on stdlib/Logging/docs/src/index.md:
```diff
@@ -0,0 +1,222 @@
+composing functionality from multiple modules.)
+
+The global logger may be set with [`global_logger`](@ref), and task-local
+loggers controlled using [`with_logger`](@ref).  Newly-spawned tasks inherit
```
Don't hyphenate compounds for an adverb ending in "ly" (google it)…

KristofferC commented:
Is there a way to just write text to a logger? Basically just want to use the stream of the logger but completely format it myself.

c42f commented:
@KristofferC logging produces a stream of *log events*, each of which has the structure described in the documentation.  You can encode formatting hints in this structure if you so please, but not all loggers will understand such hints unless there's a common convention.

Providing precise formatting control is problematic because it tightly couples the structure of events to the display capabilities of the logger which can vary widely depending on the environment (console vs juno vs ijulia for example). So you'd like to provide semantic structure in the message which can be interpreted according to the display capabilities which are present. My current best solution for rich formatting of log messages is to document that message strings should be encoded as markdown. This doesn't cover all cases (eg, color) but I think it's a practical compromise for now. I have a branch I was working on to make `ConsoleLogger` take markdown seriously, I just got a bit bogged down in refactoring `Base.Markdown` to make it actually work.

stevengj commented:
@c42f, do messages have to be strings?  If they are opaque objects that simply have appropriate `show` methods (much like documentation objects), then in principle they could have any kind of rich formatting on appropriate output devices.

c42f commented:
@stevengj there's no fundamental reason that messages need to be strings.  I feel like having strings be interpreted as markdown is a nice default which is consistent with docstrings. But using show would probably work out nicely. Perhaps something like `Logging.showmessage` with the default implementations:

```julia
showmessage(io, mime, msg) = show(io, mime, msg)
showmessage(io, mime, msg::AbstractString) = show(io, mime, Markdown.parse(msg))
```
?

stevengj commented:
@c42f, I'm skeptical that you should define a `showmessage` function.  I would think that:

* If the caller wants to pass a markdown-formatted string, they should use `md"...."` to create a `Markdown.MD` object.

* The current Logger be much like a `display` function: it should know what kinds of MIME types it can handle, call `mimewritable(mime, x)` to find out whether the message supports those MIME types, and then call `show(io, mime, x)` to output it (defaulting to text/plain if no other supported MIME types are found).

maleadt commented:
Cool, docs :-)

How would I define a custom log level? I tried:

```julia
import Base.CoreLogging: @_sourceinfo, logmsg_code, Debug
const Trace = Debug-1000
macro trace(message, exs...) logmsg_code((@_sourceinfo)..., :Trace, message, exs...) end
```

But that gives some hygiene errors on code returned by `logmsg_code`.
Or isn't this intended to be extensible yet?

Also, the `with_logger`/`global_logger` approach seems application oriented. Any suggestions/guidelines for library developers? Say, if I replace logging in my package with the new logging macros, and a user encounters some error, how do I instruct him to ramp up the log level for my package (ie. not just replacing the global logger, which is also pretty user-unfriendly)?

c42f commented:
> If the caller wants to pass a markdown-formatted string, they should use md"...." to create a Markdown.MD object.

I tried this in an earlier prototype and I didn't care for it.

Just as for docstrings, I think the simplicity and consistency of having an opinionated default makes sense here. In this system the message is "for humans" and the structured key-value pairs is a way to pass auxiliary machine readable data. So having strings which support simple formatting by convention is a win in my opinion.  If we were to advertise in the documentation that people can just put an `md` in front of the message if they feel like it, we'll probably just get less consistency and less readability across the ecosystem as a whole.  Coming back to docstrings again, making them markdown was an opinionated design choice, but I think it's worked out really well and I think there's a close analogy here.

stevengj commented:
@c42f,  in the documentation, the strings are passed through the markdown parser by default, but you can still pass any other kind of object that you want, and it uses the `show` machinery for rendering.   You aren't *prevented* from using non-Markdown strings.

c42f commented:
@stevengj yes, what I'd like to achieve is the same as the documentation system: pass strings through the markdown parser by default, and do "something sensible" with other types.  How about we change

> The *message* is a human readable string describing the event in markdown format.

to

> The *message* is an object describing the event. By convention `AbstractString`s passed as messages are assumed to be in markdown format. Other types will be displayed using `show(io,mime,obj)` according to the display capabilities of the installed logger.

c42f commented:
@maleadt good questions, mostly the answer to those is "it's not implemented yet, but yes we should have a thing there". Perhaps I should just remove mention of custom log levels for now and we can make sure they work smoothly later.  At the moment all you can do is define a custom level and use the generic `@logmsg` with that level (as in the tests).

I definitely agree that a `@trace` level is warranted.  My feeling is that it should behave a little differently from the other logging macros because for tracing, efficient filtering based on a transparent label is critical. Perhaps something like:

```
@trace :event_label a=1 b=2
```

I guess the event label goes into the `group` variable for ease of filtering.

We should learn as much as possible from the solaris dtrace system which seems super well thought out http://dtrace.org/guide/preface.html#preface.

For now I'd like to punt on the details of tracing and get this merged!

maleadt committed:
Add docs on #26265.

maleadt commented:
Rebased and added a snippet on #26265.

Anybody opposed to merging this? All review comments seem addressed, and the remaining questions are on the design of the logging system. I don't think these should hold back documentation on code that has been merged anyway.

KristofferC commented:
+1 for merging, this is `inf` better than what we currently have :)

stevengj commented:
Would like to see the clarification/fix on requiring markdown suggested by @c42f above, but otherwise +1.

maleadt committed:
Clarify handling of message argument.

[ci skip]

c42f commented:
Thanks a heap @maleadt for following this up, and sorry for being absent from the tail end of this discussion. (I'm in a new job and feeling both distracted by that and also burned out in keeping up with many open source loose ends. Bleh :-/ )
